### PR TITLE
fix synchronization in QueueingThreadPoolExecutor

### LINKED ISSUE #5031: No events received from ESH
We are currently experiencing problems with receiving all kinds of events from ESH. It looks like for certain times we do not receive any events at all from ESH. If this happens, only events from ESH are affected, while at the same time we still receive events when the OSGI EventAdmin is used directly (we have some custom events in our own bundles, where the EventAdmin is used directly).

We have seen the issues from last December (#4591 and other issues referenced there) and think this could be related, but we do not see any log messages about the EventAdmin blacklisting certain services here. So we cannot say if there is a connection, but it seems likely.

We first observed this behaviour with the 0.9.0 release version.

### PR DESCRIPTION
**Pull request by sjka**
...as before it could end up with queued tasks and no monitoring thread
if tasks were added right at the moment when the monitoring thread
decided to  terminate after there were no new tasks queued withing
two seconds.

fixes #5031
Signed-off-by: Simon Kaufmann <simon.kfm@googlemail.com>

### CODE REVIEW

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/QueueingThreadPoolExecutor.java [hunk 1]
```diff
@@ -179,7 +192,15 @@ public void run() {
                                 logger.trace("Executing queued task of thread pool '{}'.", threadPoolName);
                                 QueueingThreadPoolExecutor.super.execute(runnable);
                             } else {
-                                break;
+                                try {
+                                    lock.writeLock().lock();
```
Comment (htreu) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/QueueingThreadPoolExecutor.java]: I'm not 100% sure but shouldn't the readLock also be locked here?
Szenario: writeLock aquired here, then the Thread is paused. Switching to another Thread trying to add to the queue (line 166ff): readLock can be aquired, task added to queue, condition in line 119 is still false. When this thread continues the queuedTread is set to null and the queue again is stuck.

wdyt?

Comment (sjka) [reply bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/QueueingThreadPoolExecutor.java]: No. See https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/locks/ReentrantReadWriteLock.html

TL;DR: The locks are interconnected - while many threads can have a read-lock at the same time, the write-lock is exclusive, i.e. only one thread can have it at a time and no read-locks are possible at the same time. That's the whole clue of the ReadWriteLocks.

**Comment by maggu2810** [comment]
Can you please explain me a detail about the `QueueingThreadPoolExecutor`?

* The `EventHandler.dispatchESHEvent` method build a new `InvocationHandlerAsync` object (by calling the `build()` method and the use of `withAsync()`).
* The `receive` method is called, so the `invoke` method of the `InvocationHandlerAsync` object reference is called.
* The `SafeCallManagerImpl.enqueue` method is called.
* The `SafeCallManagerImpl.trigger` method is called.
* The `submit` method of the scheduler is called.
* `AbstractExecutorService.submit` calls the member function `execute`
* The `QueueingThreadPoolExecutor.execute` method checks if the `taskQueue` is empty and so calls (if empty) `ThreadPoolExecutor.execute` (`super.execute`)
* The `ThreadPoolExecutor` implementation itself is using the `workQueue` itself that has been supplied by the constructor.

The `workQueue` used by the `ThreadPoolExecutor` is set on construction.
The `QueueingTreadPoolExecutor` uses `new SynchronousQueue<Runnable>()`.
This one differs from the `taskQueue` used by the `QueueingTreadPoolExecutor` itself (`new LinkedTransferQueue<>()`).
Has it been implemented by intention that `workQueue != taskQueue`?
Makes the check in `QueueingTreadPoolExecutor.execute` sense?
Calling `QueueingTreadPoolExecutor.execute` multiple times will not change `taskQueue` at all.

**Comment by maggu2810** [comment]
Ah, got it. The important point is the `QueueingRejectionHandler.rejectedExecution` method.

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/QueueingThreadPoolExecutor.java (hunk 1 above)
Comment (htreu) [reply bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/QueueingThreadPoolExecutor.java]: Nice, thanks for the short explanation.
